File: backend/schemas/advice.py
# backend/schemas/advice.py
from __future__ import annotations

from typing import List, Optional, Literal

from pydantic import BaseModel, Field


# -------------------------
# Core types for the unified Advice payload
# -------------------------


AdviceStatus = Literal["ok", "disabled", "unsupported", "not_found", "timeout", "error"]


class StrategyPart(BaseModel):
    """
    Single segment of a strategy distribution.

    Each entry represents an action label and the fraction of time that action
    should be taken in this spot.  The trainer frontend maps these into
    percentage widths for the strategy bar.
    """

    action: str = Field(
        ...,
        description="Bucket/action label, e.g. 'fold', 'call', '2.5x', 'jam'.",
    )
    weight: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Probability weight in [0,1] for this action.",
    )


class AdviceMeta(BaseModel):
    """
    Minimal, street-agnostic description of the decision.

    This is the part of the payload that tells the UI where the advice applies
    (street, hero seat, number of players) and how it was produced.
    """

    street: Literal["preflop", "flop", "turn", "river", "showdown", "unknown"] = Field(
        ...,
        description="Street at which the decision occurs.",
    )
    n_players: int = Field(
        ...,
        ge=0,
        description="Number of players still in the hand at this decision.",
    )
    hero_seat: int = Field(
        ...,
        ge=0,
        description="Seat index of the hero (human) player.",
    )
    source: str = Field(
        ...,
        description=(
            "Origin of the advice, e.g. 'chart', 'equity', 'rule', 'mixed', "
            "or another backend-specific label."
        ),
    )


class AdviceRecommendation(BaseModel):
    """
    Recommended action and optional strategy distribution.

    The 'bucket' identifies the primary recommendation.  The optional
    'strategy_bar' provides a more detailed mixed strategy that can be rendered
    as a bar chart on the frontend.
    """

    bucket: str = Field(
        ...,
        description=(
            "Primary recommended bucket, e.g. 'fold', 'call', 'check', "
            "'2.5x', '2.5xR', 'jam'."
        ),
    )
    strategy_bar: Optional[List[StrategyPart]] = Field(
        default=None,
        description="Optional list of (action, weight) entries for a mixed strategy.",
    )


class AdviceEquityPlayer(BaseModel):
    """
    Per-player equity entry.

    Used when the coach computes equities for multiple players (hero + villains).
    """

    seat: int = Field(
        ...,
        ge=0,
        description="Seat index for this player.",
    )
    equity: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Normalized equity for this player in [0,1].",
    )


class AdviceEquity(BaseModel):
    """
    Optional equity block attached to advice.

    This mirrors the essential parts of /api/equity responses but is adapted for
    per-decision coaching (hero-centric).
    """

    backend: Optional[str] = Field(
        default=None,
        description="Name of the equity backend used (e.g. 'ompeval').",
    )
    mode: Optional[Literal["hands", "ranges"]] = Field(
        default=None,
        description="Equity mode: 'hands' for fixed hands, 'ranges' for ranges.",
    )
    hero: Optional[float] = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description="Hero's equity in [0,1] against the current field.",
    )
    players: Optional[List[AdviceEquityPlayer]] = Field(
        default=None,
        description="Optional list of per-player equities (including hero).",
    )
    vs_field: Optional[float] = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description="Hero equity vs everyone else combined, when available.",
    )
    exact: Optional[bool] = Field(
        default=None,
        description="Whether the result is exact enumeration (True) or Monte Carlo (False).",
    )
    iters: Optional[int] = Field(
        default=None,
        ge=0,
        description="Effective number of Monte Carlo iterations (if applicable).",
    )


class AdviceThresholds(BaseModel):
    """
    Optional threshold hints used in the rationale.

    These values help the UI and logs explain *why* a recommendation was made
    (e.g. comparing hero equity to required pot odds).
    """

    pot_odds: Optional[float] = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description="Required equity in [0,1] to continue (e.g. call) given the current price.",
    )
    spr: Optional[float] = Field(
        default=None,
        ge=0.0,
        description="Stack-to-pot ratio at this decision (hero effective stack / pot).",
    )


class AdviceV1Model(BaseModel):
    """
    Canonical shape for coach advice responses (version 1).

    This model backs the /api/coach/advice endpoint and any snapshots attached
    to exports.  All-streets, all-player-counts advice should conform to this
    schema, with optional blocks omitted when not applicable.

    See docs/COACH-ADVICE-PAYLOAD.md for the narrative specification.
    """

    # Use Literal[1] to make the version effectively constant while keeping
    # mypy and pydantic happy.
    version: Literal[1] = Field(
        1,
        description="Advice payload version.  Current version is 1.",
    )
    status: AdviceStatus = Field(
        ...,
        description=(
            "'ok' when advice is actionable; other values indicate disabled, "
            "unsupported or error states."
        ),
    )
    meta: AdviceMeta = Field(
        ...,
        description="Street- and seat-level context describing where this advice applies.",
    )
    recommendation: Optional[AdviceRecommendation] = Field(
        default=None,
        description="Recommended bucket and optional mixed strategy.",
    )
    equity: Optional[AdviceEquity] = Field(
        default=None,
        description="Optional hero/players equity information backing the advice.",
    )
    thresholds: Optional[AdviceThresholds] = Field(
        default=None,
        description="Optional pot-odds / SPR thresholds used in the rationale.",
    )
    rationale: Optional[str] = Field(
        default=None,
        description="Human-readable explanation of the recommendation.",
    )

    class Config:
        # Be tolerant of extra fields from older/newer backends.
        extra = "ignore"

File: frontend/types/advice.ts
// frontend/types/advice.ts
// Types for the unified Advice V1 payload returned by /api/coach/advice.
//
// This is shaped to match the real backend JSON, e.g.:
//
// {
//   "version": 1,
//   "status": "ok",
//   "meta": { "street": "flop", "n_players": 2, "hero_seat": 0, "source": "equity" },
//   "recommendation": { "bucket": "2.2x", "strategy_bar": [ { "action": "2.2x", "weight": 1 } ] },
//   "equity": { "backend": null, "mode": null, "hero": 0.63705, ... },
//   "thresholds": { "pot_odds": null, "spr": null, "ev_hint": null },
//   "rationale": "Hero equity ≈ ..."
// }

export type AdviceStatus =
  | 'ok'
  | 'error'
  | 'skipped'
  | 'unavailable'
  | 'disabled'
  | 'unsupported';

/**
 * High-level source label, as used by the backend in `meta.source`
 * (e.g. "equity", "solver", "chart", "noop"...).
 *
 * This is kept intentionally loose so dev tooling can treat it as a
 * simple human-readable string.
 */
export type AdviceSource = string;

/** Street literal set, matching backend semantics. */
export type AdviceStreet =
  | 'preflop'
  | 'flop'
  | 'turn'
  | 'river'
  | 'showdown'
  | 'unknown';

/** Hero’s logical table position. */
export type HeroPosition =
  | 'BTN'
  | 'SB'
  | 'BB'
  | 'UTG'
  | 'HJ'
  | 'CO'
  | 'unknown';

/** Mixed strategy over action labels. Keys are bucket labels. */
export type ActionMix = Record<string, number>;

/**
 * Generic strategy-part shape used by legacy coach types.
 * `types/coach.ts` re-exports this as `AdviceStrategyPart`.
 */
export interface StrategyPart {
  action: string;
  weight: number;
}

/** Alias kept for clarity in newer code. */
export type AdviceStrategyPart = StrategyPart;

/**
 * The coach’s recommendation for this spot.
 *
 * The backend currently sends:
 *   recommendation.bucket        – canonical chosen bucket/action id
 *   recommendation.strategy_bar  – [{ action, weight }, ...]
 *
 * Early prototypes used:
 *   recommendation.primary_action
 *   recommendation.action_mix
 *
 * We support both for compatibility.
 */
export interface AdviceRecommendation {
  /** Canonical bucket/action id, e.g. "fold", "call", "2.5x", "jam". */
  bucket?: string;

  /** Backwards-compatible alias used in early drafts. */
  primary_action?: string;

  /** Canonical strategy bar from backend. */
  strategy_bar?: AdviceStrategyPart[];

  /** Optional map-form mix used by older code. */
  action_mix?: ActionMix;

  /** Optional sizing hint (rarely populated). */
  sizing_hint?: string | null;
}

/**
 * Per-seat equity entry (multiway support).
 * Not always present; often only hero is returned.
 */
export interface AdviceEquityPlayer {
  seat: number;
  equity: number;
}

/**
 * Equity block from the backend.
 *
 * In your current payload this looks like:
 *   "equity": {
 *     "backend": null,
 *     "mode": null,
 *     "hero": 0.63705,
 *     "players": null,
 *     "vs_field": null,
 *     "exact": null,
 *     "iters": null
 *   }
 *
 * We also expose a few optional, more "UI-friendly" aliases so newer
 * components (like the DecisionHelpOverlay) can attach to them without
 * breaking older callers:
 *
 *   hero_vs_villain_equity  – alias for `hero`
 *   pot_odds                – often comes from thresholds; may be copied over
 *   min_equity_to_call      – same as above
 */
export interface AdviceEquity {
  backend: string | null;
  mode: string | null;
  hero: number | null;
  players: AdviceEquityPlayer[] | null;
  vs_field: unknown | null;
  exact: boolean | null;
  iters: number | null;

  /** Optional text comment, if backend or adapter provides one. */
  comment?: string | null;

  /** Optional normalised aliases used by newer UIs. */
  hero_vs_villain_equity?: number | null;
  pot_odds?: number | null;
  min_equity_to_call?: number | null;
}

/**
 * Thresholds / pot-odds hints attached to the equity.
 */
export interface AdviceThresholds {
  /** Pot odds required equity to continue (0..1). */
  pot_odds?: number | null;

  /** Minimum equity required to call given the price (0..1). */
  min_equity_to_call?: number | null;

  /** Stack-to-pot ratio, if provided. */
  spr?: number | null;

  /** Optional EV hint string from backend. */
  ev_hint?: string | null;
}

/**
 * Optional richer context block. Not all fields will necessarily be
 * populated yet; everything is optional on purpose.
 */
export interface AdviceContext {
  street?: AdviceStreet | string;
  hero_position?: HeroPosition | string;
  hero_seat?: number;
  hero_cards?: string[] | null;
  board?: string[];
  pot_size?: number;
  to_call?: number;
  stack_effective?: number | null;
  allowed_buckets?: string[];
}

/**
 * Light-weight metadata; superseded by `context` when present.
 * This mirrors the `meta` object you see in the JSON.
 */
export interface AdviceMeta {
  street?: AdviceStreet | string;
  n_players?: number;
  hero_seat?: number;
  /** High-level backend source, e.g. "equity", "solver", "chart". */
  source?: AdviceSource;
}

/**
 * Canonical Advice V1 envelope from the backend.
 *
 * This is what `/api/coach/advice` returns today.
 */
export interface AdvicePayloadV1 {
  version: 1;
  status: AdviceStatus;

  /** Simple metadata; new code should prefer `context` where available. */
  meta?: AdviceMeta;

  /** Optional richer, UI-focused context (future-facing). */
  context?: AdviceContext;

  /**
   * Optional explicit source object or label. Most current backends
   * only populate meta.source, but this gives UI code a stable place
   * to hang a human-readable "Source:" badge if we add it later.
   */
  source?: AdviceSource | null;

  recommendation?: AdviceRecommendation;
  equity?: AdviceEquity;
  thresholds?: AdviceThresholds;

  /** Optional free-form explanation string. */
  rationale?: string | null;
}

/** Convenience aliases used throughout the frontend. */
export type Advice = AdvicePayloadV1;
export type AdviceV1 = AdvicePayloadV1;
